h2o-core/src/main/java/water/rapids/ast/prims/assign/AstRectangleAssign.java

```diff
@@ -284,23 +285,78 @@ void mapChunkSlice(Chunk[] cs, int chkOffset) {
     }
   }
 
+  private boolean isScalarCompatible(Object scalar, Vec v) {
+    if (scalar == null)
+      return true;
+    else if (scalar instanceof Number)
+      return v.get_type() == Vec.T_NUM || v.get_type() == Vec.T_TIME;
+    else if (scalar instanceof String) {
+      if (v.get_type() == Vec.T_CAT) {
+        for (String f: v.domain()) if (f.equals(scalar)) return true;
+        return false;
+      } else
+        return v.get_type() == Vec.T_STR || (v.get_type() == Vec.T_UUID);
+    } else
+      return false;
+  }
+
   // Boolean assignment with a scalar
-  private void assign_frame_scalar(Frame dst, final int[] cols, Frame rows, final double src, Session ses) {
-    // TODO: COW without materializing vec and depending on assign_frame_frame
-    Frame src2 = new MRTask() {
-      @Override
-      public void map(Chunk[] cs, NewChunk[] ncs) {
-        Chunk bool = cs[cs.length - 1];
-        for (int i = 0; i < cs[0]._len; ++i) {
-          int nc = 0;
-          if (bool.at8(i) == 1)
-            for (int ignored : cols) ncs[nc++].addNum(src);
-          else
-            for (int c : cols) ncs[nc++].addNum(cs[c].atd(i));
-        }
+  private void assign_frame_scalar(Frame dst, int[] cols, Frame rows, Object src, Session ses) {
+    // Bulk assign a numeric constant over a frame. Directly set columns without checking target type
+    // assuming the user just wants to overwrite everything: Copy-On-Write optimization happens here on the apply() exit.
+    // Note: this skips "scalar to Vec" compatibility check because the whole Vec is overwritten
+    Vec bool = rows.vec(0);
+    if (bool.isConst() && ((int) bool.min() == 1) && (src instanceof Number)) {
+      Vec anyVec = dst.anyVec();
+      assert anyVec != null;
+      Vec vsrc = anyVec.makeCon((double) src);
+      for (int col : cols)
+        dst.replace(col, vsrc);
+      if (dst._key != null) DKV.put(dst);
+      return;
+    }
+
+    // Make sure the scalar value is compatible with the target vector
+    for (int col: cols) {
+      if (! isScalarCompatible(src, dst.vec(col))) {
+        throw new IllegalArgumentException("Cannot assign value " + src + " into a vector of type " + dst.vec(col).get_type_str() + ".");
       }
-    }.doAll(cols.length, Vec.T_NUM, new Frame(dst).add(rows)).outputFrame();
-    assign_frame_frame(dst, cols, new AstNumList(0, dst.numRows()), src2, ses);
+    }
+
+    Vec[] vecs = ses.copyOnWrite(dst, cols);
+    Vec[] vecs2 = new Vec[cols.length]; // Just the selected columns get updated
+    for (int i = 0; i < cols.length; i++)
+      vecs2[i] = vecs[cols[i]];
+
+    new ConditionalAssignTask(src).doAssign(vecs2, rows.vec(0));
+  }
+
+  private static class ConditionalAssignTask extends MRTask<ConditionalAssignTask> {
+
+    final Serializable _value;
```

Comment:
Does this work? Our serialization now works with Serializable?
It might be better to make the ValueSetter Iced and pass the ValueSetter itself (and pass it a chunk in set).